### DISCUSSION TITLE
Prevent DDOS attacks while waiting for client connection

### DISCUSSION THREAD

**Discussion by falkoschindler** [zauberzeug/nicegui#1727] [Ideas / Feature Requests]
As discussed in #1695, there is the possibility for a DDOS attack: If the client requests a page but doesn't connect via websocket, the server keeps waiting for 3 seconds. In this time window there could be thousands of requests from the same client, causing acquired resources to accumulate.

We could introduce a parameter for `ui.page` like `max_connections_per_ip: int = 100` to limit the number of simultaneous requests per client.

Besides that it might be useful to make the client IP accessible _before_ the websocket connection is established.

**Comment by DoPaminexL** [comment]
Can you provide an example to give me a reference? I really need this use of this max_connections_per_ip

**Reply by falkoschindler** [reply]
@DoPaminexL I'm not sure what you mean. `max_connections_per_ip` is not implemented yet.

To further illustrate the problem:

```py
@ui.page("/")
async def index(client: Client):
    await client.connected()
    connect_ip = client.ip
    print(connect_ip)
```

To get the IP from the client (for whatever reason) we currently need to await the `client.connected()` method, which has a default timeout of 3 seconds. If someone just requests the page, e.g. via `curl`, the socket connection is never established, but the client object is held for 3 seconds. This could be misused to stress the server.

**Comment by rodja** [comment]
I've just created pull request #2399 to make the ip property available before websocket connection is established.
